Bottom: c18708df17502f44ce765d8e0d2cf96d59e9d644
Top:    8fdab208b53f375384755c44c073ecd00f356f0f
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-24 22:44:07 +0000

DEBUG STUFF - looking into bad polygons


---

diff --git a/src/polygon.c b/src/polygon.c
index c1f78c1..3b94ae2 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -695,6 +695,9 @@ ArcPoly (ArcType * a, Coord thick)
   return ArcPolyNoIntersect (a, thick);
 }
 
+void pline_dump (VNODE * v);
+void poly_dump (POLYAREA * p);
+
 POLYAREA *
 LinePoly (LineType * L, Coord thick)
 {
@@ -763,6 +766,8 @@ LinePoly (LineType * L, Coord thick)
   if (!(np = ContourToPoly (contour)))
     return NULL;
 
+  poly_dump (np);
+
   return np;
 }
 
@@ -1126,10 +1131,28 @@ line_sub_callback (const BoxType * b, void *cl)
   if (!(np = LinePoly (line, line->Thickness + line->Clearance)))
     longjmp (info->env, 1);
 
+  if (info->batch_size == 2)
+    {
+      printf ("Batch no %i, input:\n",info->batch_size);
+      poly_dump (info->accumulate);
+      printf ("---- subtract from it:\n");
+      poly_dump (np);
+    }
+
   poly_Boolean_free (info->accumulate, np, &merged, PBO_UNITE);
+
+  if (info->batch_size == 2)
+    {
+      printf ("Ouptut:\n");
+      poly_dump (merged);
+      printf ("-----------\n");
+    }
+
   info->accumulate = merged;
   info->batch_size ++;
 
+  pcb_printf ("Line %mm, %mm - %mm, %mm\n", line->Point1.X, line->Point1.Y, line->Point2.X, line->Point2.Y);
+
   if (info->batch_size == SUBTRACT_LINE_BATCH_SIZE)
     subtract_accumulated (info, polygon);
 
@@ -1201,7 +1224,8 @@ clearPoly (DataType *Data, LayerType *Layer, PolygonType * polygon,
         r +=
           r_search (layer->line_tree, &region, NULL, line_sub_callback,
                     &info);
-        subtract_accumulated (&info, polygon);
+//        subtract_accumulated (&info, polygon);
+      polygon->Clipped = info.accumulate;
         r +=
           r_search (layer->arc_tree, &region, NULL, arc_sub_callback, &info);
 	r +=
@@ -1210,7 +1234,7 @@ clearPoly (DataType *Data, LayerType *Layer, PolygonType * polygon,
       END_LOOP;
       r += r_search (Data->via_tree, &region, NULL, pin_sub_callback, &info);
       r += r_search (Data->pin_tree, &region, NULL, pin_sub_callback, &info);
-      subtract_accumulated (&info, polygon);
+//      subtract_accumulated (&info, polygon);
     }
   polygon->NoHolesValid = 0;
   return r;
diff --git a/src/polygon1.c b/src/polygon1.c
index 950ea0d..82e64dc 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -109,12 +109,12 @@ int vect_inters2 (Vector A, Vector B, Vector C, Vector D, Vector S1,
 
 #define error(code)  longjmp(*(e), code)
 
-#undef DEBUG_LABEL
-#undef DEBUG_ALL_LABELS
-#undef DEBUG_JUMP
-#undef DEBUG_GATHER
-#undef DEBUG_ANGLE
-#undef DEBUG
+#define DEBUG_LABEL
+#define DEBUG_ALL_LABELS
+#define DEBUG_JUMP
+#define DEBUG_GATHER
+#define DEBUG_ANGLE
+#define DEBUG
 #ifdef DEBUG
 #define DEBUGP(...) pcb_fprintf(stderr, ## __VA_ARGS__)
 #else
@@ -154,7 +154,7 @@ pline_dump (VNODE * v)
   do
     {
       n = NEXT_VERTEX(v);
-      pcb_fprintf (stderr, "Line [%#mS %#mS %#mS %#mS 10 10 \"%s\"]\n",
+      pcb_fprintf (stderr, "Line [%mn %mn %mn %mn 10 10 \"%s\"]\n",
 	       v->point[0], v->point[1],
 	       n->point[0], n->point[1], theState (v));
     }
@@ -277,8 +277,9 @@ new_descriptor (VNODE * a, char poly, char side)
   l->angle = ang;
   assert (ang >= 0.0 && ang <= 4.0);
 #ifdef DEBUG_ANGLE
-  DEBUGP ("node on %c at %#mD assigned angle %g on side %c\n", poly,
-	  a->point[0], a->point[1], ang, side);
+  DEBUGP ("node on %c at %mn, %mn assigned angle %g on side %c\n", poly,
+	  a->point[0], a->point[1], ang * 10000., side);
+  DEBUGP ("That angle would be %f\n", atan2 (v[1], v[0]) * 180. / M_PI);
 #endif
   return l;
 }
@@ -1182,7 +1183,7 @@ print_labels (PLINE * a)
 
   do
     {
-      DEBUGP ("%#mD->%#mD labeled %s\n",
+      DEBUGP ("%mm, %mm -> %mm, %mm labeled %s\n",
               EDGE_BACKWARD_VERTEX (e)->point[0], EDGE_BACKWARD_VERTEX (e)->point[1],
                EDGE_FORWARD_VERTEX (e)->point[0],  EDGE_FORWARD_VERTEX (e)->point[1], theState (e));
     }
@@ -1686,7 +1687,7 @@ jump (VNODE **curv, DIRECTION *cdir, J_Rule j_rule)
       return TRUE;
     }
 #ifdef DEBUG_JUMP
-  DEBUGP ("jump entering node at %$mD\n", (*curv)->point[0], (*curv)->point[1]);
+  DEBUGP ("jump entering node at %mm, %mm\n", (*curv)->point[0], (*curv)->point[1]);
 #endif
   /* Pick the descriptor of the edge we came into this vertex with, then spin (anti?)clock-wise one edge descriptor */
   if (*cdir == FORW)
@@ -1705,7 +1706,7 @@ jump (VNODE **curv, DIRECTION *cdir, J_Rule j_rule)
 	      (d->side == 'P' && newone == BACKW))
 	    {
 #ifdef DEBUG_JUMP
-	      DEBUGP ("jump leaving node at %#mD\n",
+	      DEBUGP ("jump leaving node at %mm, %mm\n",
 	              EDGE_DIRECTION_VERTEX (e, newone)->point[0], EDGE_DIRECTION_VERTEX (e, newone)->point[1]);
 #endif
 	      *curv = d->parent;
@@ -1745,7 +1746,7 @@ Gather (VNODE *startv, PLINE **result, J_Rule j_rule, DIRECTION initdir)
 	  poly_InclVertex (PREV_VERTEX (&(*result)->head), newn);
 	}
 #ifdef DEBUG_GATHER
-      DEBUGP ("gather vertex at %#mD\n", curv->point[0], curv->point[1]);
+      DEBUGP ("gather vertex at %mm, %mm\n", curv->point[0], curv->point[1]);
 #endif
       /* Now mark the edge as included.  */
       newn = VERTEX_DIRECTION_EDGE (curv, dir);
